<!--
  1 failure:  Tests have failed...
  0 warning: 
  
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| android.Titanium.Geolocation | reverseGeocoder | 2.481 | AssertionError: expected false to equal true |
| android.Titanium.Network.HTTPClient | clearCookieUnaffectedCheck | 60.128 | Error: timeout of 60000ms exceeded |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

--- a/iphone/Classes/KrollContext.h
+++ b/iphone/Classes/KrollContext.h
@@ -198,10 +198,10 @@ KrollContext *GetKrollContext(TiContextRef context);
 /**
  * Map of timer identifiers and the underlying native NSTimer.
  */
-@property (nonatomic, strong) NSMutableDictionary<NSNumber *, NSTimer *> *timers;
+@property (nonatomic, strong) NSMapTable<NSNumber *, NSTimer *> *timers;
 
 /**
- * Initailizes the timer manager in the given JS context. Exposes the global set/clear
+ * Initializes the timer manager in the given JS context. Exposes the global set/clear
  * functions for creating and clearing intervals/timeouts.
  *
  * @param context The JSContext where timer function should be made available to.

--- a/iphone/Classes/KrollContext.m
+++ b/iphone/Classes/KrollContext.m
@@ -1622,7 +1622,8 @@ static TiValueRef StringFormatDecimalCallback(TiContextRef jsContext, TiObjectRe
 
 - (void)invalidateAllTimers
 {
-  for (NSNumber *timerIdentifier in self.timers) {
+  NSArray<NSNumber *> *keys = [[self.timers keyEnumerator] allObjects];
+  for (NSNumber *timerIdentifier in keys) {
     [self clearIntervalWithIdentifier:timerIdentifier.unsignedIntegerValue];
   }
 }

--- a/iphone/Classes/KrollContext.m
+++ b/iphone/Classes/KrollContext.m
@@ -1576,23 +1576,34 @@ static TiValueRef StringFormatDecimalCallback(TiContextRef jsContext, TiObjectRe
   }
 
   self.nextTimerIdentifier = 0;
-  self.timers = [NSMutableDictionary new];
+  self.timers = [NSMapTable strongToWeakObjectsMapTable];
 
-  NSUInteger (^setInterval)(void) = ^() {
-    return [self setIntervalFromArguments:JSContext.currentArguments shouldRepeat:YES];
+  NSUInteger (^setInterval)(JSValue *, double) = ^(JSValue *callback, double interval) {
+    return [self setInterval:interval withCallback:callback shouldRepeat:YES];
   };
   context[@"setInterval"] = setInterval;
 
-  NSUInteger (^setTimeout)(void) = ^() {
-    return [self setIntervalFromArguments:JSContext.currentArguments shouldRepeat:NO];
+  NSUInteger (^setTimeout)(JSValue *, double) = ^(JSValue *callback, double interval) {
+    return [self setInterval:interval withCallback:callback shouldRepeat:NO];
   };
   context[@"setTimeout"] = setTimeout;
 
-  void (^clearInterval)(JSValue *) = ^(JSValue *value) {
-    return [self clearIntervalWithIdentifier:value.toInt32];
+  void (^clearInterval)(NSUInteger) = ^(NSUInteger value) {
+    return [self clearIntervalWithIdentifier:value];
   };
   context[@"clearInterval"] = clearInterval;
   context[@"clearTimeout"] = clearInterval;
+
+  // This is more useful than just in timers, should be registered in some better place like KrollBridge?
+  [context setExceptionHandler:^(JSContext *context, JSValue *exception) {
+    id exc;
+    if ([exception isObject]) {
+      exc = [exception toObject]; // Hope it becomes an NSDictionary?
+    } else {
+      exc = [exception toString];
+    }
+    [[TiExceptionHandler defaultExceptionHandler] reportScriptError:[TiUtils scriptErrorValue:exc]];
+  }];
 
   return self;
 }
@@ -1600,6 +1611,7 @@ static TiValueRef StringFormatDecimalCallback(TiContextRef jsContext, TiObjectRe
 - (void)dealloc
 {
   if (self.timers != nil) {
+    [self invalidateAllTimers];
     [self.timers removeAllObjects];
     [self.timers release];
     self.timers = nil;
@@ -1610,36 +1622,54 @@ static TiValueRef StringFormatDecimalCallback(TiContextRef jsContext, TiObjectRe
 
 - (void)invalidateAllTimers
 {
-  for (NSNumber *timerIdentifier in self.timers.allKeys) {
+  for (NSNumber *timerIdentifier in self.timers) {
     [self clearIntervalWithIdentifier:timerIdentifier.unsignedIntegerValue];
   }
 }
 
-- (NSUInteger)setIntervalFromArguments:(NSArray<JSValue *> *)arguments shouldRepeat:(BOOL)shouldRepeat
+- (NSUInteger)setInterval:(double)interval withCallback:(JSValue *)callback shouldRepeat:(BOOL)shouldRepeat
 {
-  NSMutableArray *callbackArgs = [arguments.mutableCopy autorelease];
-  JSValue *callbackFunction = [callbackArgs objectAtIndex:0];
-  [callbackArgs removeObjectAtIndex:0];
-  double interval = [[callbackArgs objectAtIndex:0] toDouble] / 1000;
-  [callbackArgs removeObjectAtIndex:0];
+  // interval is optional, should default to 0 per spec, but let's enforce at least 1 ms minimum (like Node)
+  if (isnan(interval) || interval < 1) {
+    interval = 1; // defaults to 1ms
+  }
+  interval = interval / 1000.0; // convert from ms to seconds
+
+  // Handle additional arguments being passed in
+  NSMutableArray<JSValue *> *callbackArgs = [[JSContext currentArguments].mutableCopy autorelease];
+  if ([callbackArgs count] > 1) {
+    [callbackArgs removeObjectAtIndex:1]; // drop interval
+  }
   NSNumber *timerIdentifier = @(self.nextTimerIdentifier++);
-  NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:interval
-                                                   repeats:shouldRepeat
-                                                     block:^(NSTimer *_Nonnull timer) {
-                                                       [callbackFunction callWithArguments:callbackArgs];
-                                                       if (!shouldRepeat) {
-                                                         [self.timers removeObjectForKey:timerIdentifier];
-                                                       }
-                                                     }];
-  self.timers[timerIdentifier] = timer;
+  NSTimer *timer = [NSTimer timerWithTimeInterval:interval target:self selector:@selector(callJsCallback:) userInfo:callbackArgs repeats:shouldRepeat];
+  [[NSRunLoop mainRunLoop] addTimer:timer forMode:NSRunLoopCommonModes];
+
+  [self.timers setObject:timer forKey:timerIdentifier];
   return [timerIdentifier unsignedIntegerValue];
+}
+
+- (void)callJsCallback:(NSTimer *_Nonnull)timer
+{
+  NSArray<JSValue *> *args = [timer userInfo];
+  JSValue *callback = [args objectAtIndex:0];
+  NSUInteger argCount = [args count];
+  NSArray<JSValue *> *varargs = nil;
+  if (argCount > 1) {
+    varargs = [args subarrayWithRange:NSMakeRange(1, argCount - 1)];
+  }
+  [callback callWithArguments:varargs];
 }
 
 - (void)clearIntervalWithIdentifier:(NSUInteger)identifier
 {
   NSNumber *timerIdentifier = @(identifier);
-  [self.timers[timerIdentifier] invalidate];
-  [self.timers removeObjectForKey:timerIdentifier];
+  NSTimer *timer = [self.timers objectForKey:timerIdentifier];
+  if (timer != nil) {
+    if ([timer isValid]) {
+      [timer invalidate];
+    }
+    [self.timers removeObjectForKey:timerIdentifier];
+  }
 }
 
 @end

